Multus with SRIOV CNI Plugin first E2E script

lennybe opened the pull request:
Readme updated
E2E script added (Currently Mellanox NIC support only)

Commit:
Multus with SRIOV CNI Plugin first E2E script

Readme updated
E2E script added (Currently Mellanox NIC support only)

lennybe commented:
recheck

lennybe commented:
recheck

lennybe commented:
recheck

Commit:
Added kubernetes install

Mmduh-483 commented on README.md:
```diff
@@ -1,2 +1,2 @@
 # Kubernetes Tests
-CI for Kubernetes Tests
+CI for SRIOV CNI
```
It would be better to write 
CI for deploying Kubernetes with Multus, SRIOV CNI, and Device Plugin

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+export ARTIFACTS=$WORKSPACE/artifacts
+
+export PATH=/usr/local/go/bin/:/opt/cni/bin/:/home_stack/cni/cni/cnitool/$PATH
+export CNI_PATH=/opt/cni/bin/
```
Remove line 9 you are not using the `CNI_PATH` also it is the same `CNI_BIN_DIR`

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+#!/bin/bash -x
+
```
Add `set -e` so the script will fail if any of the commands failed

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+
+SRIOV_NETWORK_DEVICE_PLUGIN_REPO=${SRIOV_NETWORK_DEVICE_PLUGIN_REPO:-https://github.com/intel/sriov-network-device-plugin}
+SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH=${SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH:-remotes/origin/feature/device_selectors}
+
```
https://github.com/Mellanox/sriov_cni_ci/blob/fa105c93c8af1f9447bc1f3e839398f347ce82c2/multus_sriov_cni.sh#L11-L21
move them after all the exports

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+export LOGDIR=$WORKSPACE/logs
+export ARTIFACTS=$WORKSPACE/artifacts
+
+export PATH=/usr/local/go/bin/:/opt/cni/bin/:/home_stack/cni/cni/cnitool/$PATH
```
It should be `:$PATH` also add `kubernetes/third_party/etcd` it is needed when run kubernetes

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+function clear_vfs {
+    echo 0 > /sys/class/net/$SRIOV_INTERFACE/device/sriov_numvfs
+    echo 4 > /sys/class/net/$SRIOV_INTERFACE/device/sriov_numvfs
+}
```
Create _Vfs may be a better name, also please consider adding a parameter for the number of vfs and 4 is the default

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+}
+
+
+function install_k8s {
```
Split `install_k8s` into 2 functions  clone_k8s and run k8s

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+    go get -u github.com/tools/godep
+    go get -u github.com/cloudflare/cfssl/cmd/...
+    $GOPATH/src/k8s.io/kubernetes/hack/install-etcd.sh
+    export PATH="$GOPATH/src/k8s.io/kubernetes/third_party/etcd:${PATH}"
```
`export PATH="$GOPATH/src/k8s.io/kubernetes/third_party/etcd:${PATH}"`
This is needed always to run k8s, consider moving it to the export PATH at the start of the script

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+clear_vfs
+download_cni
+
+K8S_PID=$(install_k8s)
```
Stop previous kubernetes if exist then start a new one and remove the `kubectl delete ...` lines they will not be needed then

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+
+kubectl delete -f $WORKSPACE/sriov-network-device-plugin/deployments/sriov-crd.yaml
+kubectl create -f $WORKSPACE/sriov-network-device-plugin/deployments/sriov-crd.yaml
+pod_create
```
deploying `sriov-crd.yaml` should be after deploying multus directly

zenghui-shi replied:
Also make sure `crdnetwork.yaml` is created first, otherwise `sriov-crd.yaml` creation would fail.

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+
+    kubectl delete -f $WORKSPACE/multus-cni/images/multus-daemonset.yml
+    kubectl create -f $WORKSPACE/multus-cni/images/multus-daemonset.yml
+    sleep 60
```
` kubectl create -f $WORKSPACE/multus-cni/images/multus-daemonset.yml`
Should be after deploying k8s not before, we are not sure if there is a previous k8s runnig

Mmduh-483 commented:
Please consider creating all the artifacts in one function then deploy them

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+    pushd $WORKSPACE/sriov-cni
+    git checkout $SRIOV_CNI_BRANCH
+    git log -p -1 > $ARTIFACTS/sriov-cni.txt
+    popd
```
Where are you building the sriov-cni binary and moving it to the CNI_BIN_DIR?, a better way is to deploy the sriov-cni daemonset `kubectl create -f sriov-cni/image/sriov-cni-daemonset.yaml`

Commit:
increased number of resources

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,247 @@
+    cp /tmp/sriov_pod.yaml $ARTIFACTS
+    kubectl delete -f /tmp/sriov_pod.yaml
+    kubectl create -f /tmp/sriov_pod.yaml
+
```
the location should be in `$artifacts//sriov_pod.yaml`

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+PLUGINS_REPO=${PLUGINS_REPO:-https://github.com/containernetworking/plugins.git}
+
+SRIOV_NETWORK_DEVICE_PLUGIN_REPO=${SRIOV_NETWORK_DEVICE_PLUGIN_REPO:-https://github.com/intel/sriov-network-device-plugin}
+SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH=${SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH:-remotes/origin/feature/device_selectors}
```
Should be testing master branch now, all features from device_selectors has been merged to master.

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+    kubectl delete -f $WORKSPACE/multus-cni/images/multus-daemonset.yml
+    kubectl create -f $WORKSPACE/multus-cni/images/multus-daemonset.yml
+    sleep 60
+    cat > $CNI_CONF_DIR/00-multus.conf <<EOF
```
Is this going to overwrite the config file generated by multus-daemonset.yml? If yes, then what about passing the config file to multus pod via args `--multus-conf-file=` ?

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+        return 0
+    fi
+
+    echo "Download $MULTUS_CNI_REPO"
```
This is not needed if `multus-daemonset.yml` is used to install multus cni binary.

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+    pushd $WORKSPACE/plugins
+    git checkout $PLUGINS_BRANCH
+    git log -p -1 > $ARTIFACTS/plugins.txt
+    ./build_linux.sh
```
Copy the built CNI binaries to `$CNI_BIN_DIR` ?

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+        - sh
+        - -c
+        - |
+          ls -l /dev/infiniband /sys/class/net
```
`/dev/infiniband` may not exist as the pod is not asking for a rdma device?

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+      imagePullPolicy: IfNotPresent
+      securityContext:
+        capabilities:
+          add: ["IPC_LOCK"]
```
It seems that `test_pod` only runs `ip a` cmd to test the pod, do we need `IPC_LOCK` capability?

zenghui-shi commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+K8S_PID=$(install_k8s)
+configure_multus $MACVLAN_INTERFACE
+kill $(pgrep sriovdp)
+nohup $WORKSPACE/sriov-network-device-plugin/build/sriovdp -logtostderr 10 2>&1|tee > $LOGDIR/sriovdp.log &
```
`sriov-network-device-plugin` can also be run as daemonset, `--resource-prefix` cmd option can be used to change the resource name prefix if you're testing mellanox cards, for example:
```
./build/sriovdp --resource-prefix=mellanox.com --logtostderr 10
```
Then exposed resource name should be `mellanox.com/sriov`.

zshi-redhat reviewed (commented):
Thanks for the PR, looks pretty good!
several comments inline.

zshi-redhat commented:
Do we want to gate each PR of SR-IOV CNI and Device Plugin before it gets merged? if yes, then we may want to add support of pulling the PR and building cni & device plugin binaries on top of it.

zshi-redhat commented:
@lennybe @Mmduh-483 do you know if `local-up-cluster.sh` is able to deploy a released kubernetes version such as 1.12, 1.13 etc?

Mmduh-483 commented:
> @lennybe @Mmduh-483 do you know if `local-up-cluster.sh` is able to deploy a released kubernetes version such as 1.12, 1.13 etc by passing the release version?

first, we need to change to the release branch then run the `local-up-cluster.sh` script

Mmduh-483 commented on multus_sriov_cni.sh:
```diff
@@ -0,0 +1,249 @@
+spec:
+  restartPolicy: OnFailure
+  containers:
+    - image: mellanox/mlnx_ofed_linux-4.4-1.0.0.0-centos7.4
```
consider using alpine image since it is a very small image
`- image: alpine`

Commit:
Devided and applied comments

Applied most of the comments
Devided to separated scripts for install and test

Signed-off-by: Lenny Verkhovsky <lennyb@mellanox.com>

zenghui-shi commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+    git checkout $SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH
+    git log -p -1 > $ARTIFACTS/sriov-network-device-plugin.txt
+    make build
+    cp $WORKSPACE/sriov-network-device-plugin/build/sriovdp $CNI_BIN_DIR
```
No need to copy `sriovdp` into `$CNI_BIN_DIR`, it's not a cni binary, will be executed explicitly.

zenghui-shi commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+    git clone $SRIOV_CNI_REPO $WORKSPACE/sriov-cni
+    pushd $WORKSPACE/sriov-cni
+    git checkout $SRIOV_CNI_BRANCH
+    git log -p -1 > $ARTIFACTS/sriov-cni.txt
```
I didn't find where `sriov` cni binary is installed. shall we do it here or deploy it via `sriov-cni-daemonset.yaml`?

Mmduh-483 replied:
it would be better to use the daemonset

zenghui-shi commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+
+install_k8s
+configure_multus
+while [ ! -f $WORKSPACE/sriov-network-device-plugin/deployments/crdnetwork.yaml ]; do
```
This file will be exist once `sriov-network-device-plugin` repo is cloned.
Instead of checking the file existence, we shall create the crd by running `kubectl create -f $WORKSPACE/sriov-network-device-plugin/deployments/crdnetwork.yaml`

zenghui-shi commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+
+    echo "Configure Multus"
+    kubectl create -f $WORKSPACE/multus-cni/images/multus-daemonset.yml
+    sleep 60
```
Instead of `sleep`, perhaps we could check the ds status and if the daemonset pod is running.

zenghui-shi commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+    git checkout $PLUGINS_BRANCH
+    git log -p -1 > $ARTIFACTS/plugins.txt
+    bash ./build_linux.sh
+    cp $WORKSPACE/plugins/bin/* $CNI_BIN_DIR/
```
Do we clean up the `#CNI_BIN_DIR` in each run? if not, this may fail due to file exists.
perhaps add force copy?

lennybe commented:
recheck

lennybe commented:
recehck

lennybe commented:
recheck

Mmduh-483 commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+
+export GOROOT=${GOROOT:-/usr/local/go}
+export GOPATH=${WORKSPACE}
+export PATH=/usr/local/go/bin/:/opt/cni/bin/:$GOPATH/src/k8s.io/kubernetes/third_party/etcd:$PATH
```
no need to add `/opt/cni/bin/` to the PATH

Mmduh-483 commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+    done
+    return $k8s_pid
+}
+
```
if k8s failed to run then this process will keep running forever, make it fail after 5 mins if it didn't work

Mmduh-483 commented on multus_sriov_cni_install.sh:
```diff
@@ -0,0 +1,202 @@
+export SRIOV_NETWORK_DEVICE_PLUGIN_REPO=${SRIOV_NETWORK_DEVICE_PLUGIN_REPO:-https://github.com/intel/sriov-network-device-plugin}
+export SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH=${SRIOV_NETWORK_DEVICE_PLUGIN_BRANCH:-master}
+
+export GOROOT=${GOROOT:-/usr/local/go}
```
`GOROOT` is not used anywhere in the code

lennybe commented:
recheck

lennybe commented:
recheck

Commit:
Multus with SRIOV CNI Plugin

Signed-off-by: Lenny Verkhovsky <lennyb@mellanox.com>

lennybe reviewed (commented):
recheck

lennybe commented:
recheck

zshi-redhat commented:
> recheck

I suppose `recheck` runs some CI tests for this PR, did the test pass?

Commit:
E2E with K8S, Multus and SRIOV-CNI Plugin

lennybe commented:
recheck

Commit:
Download kubectl

lennybe commented:
recheck

Commit:
removed trailing space

Commit:
removed trailing space

lennybe commented:
recheck

lennybe commented:
recheck
